arca/backend/docker.py

```diff
@@ -0,0 +1,271 @@
+            RUN mkdir /srv/scripts/
+            COPY {branch} /srv/data/
+            {dependencies}
+            {requirements_file}
+
+            WORKDIR {workdir.resolve()}
+
+            CMD bash -i
+        """)
+
+        logger.info("Started building %s:%s", image_name, image_tag)
+        try:
+            image = self.client.images.build(
+                path=str(repo_path.parent.resolve()),
+                tag=f"{image_name}:{image_tag}",
+                dockerfile=f"docker_{branch}"
+            )
+        except BuildError as e:
+            logger.exception(e)
+            raise
+
+        logger.info("Finished building %s:%s", image_name, image_tag)
+
+        return image
+
+    def image_exists(self, image_name, image_tag):
+        return f"{image_name}:{image_tag}" in itertools.chain(*[image.tags
+                                                                for image in self.client.images.list()])
+
+    def get_or_create_environment(self, repo: str, branch: str) -> Image:
+        image_name = self.get_image_name(repo, branch)
+        image_tag = self.get_image_tag(repo, branch)
+
+        if self.image_exists(image_name, image_tag):
+            return self.client.images.get(f"{image_name}:{image_tag}")
+
+        return self.create_image(repo, branch, image_name, image_tag)
+
+    def container_running(self, image) -> Optional[Container]:
+        for container in self.client.containers.list():
+            if image == container.image:
+                return container
+        return None
+
+    def start_container(self, image) -> Container:
+        return self.client.containers.run(image, command="bash -i", detach=True, tty=True)
+
+    def tar_file(self, name, script):
+        tarstream = BytesIO()
+        tar = tarfile.TarFile(fileobj=tarstream, mode='w')
+        tarinfo = tarfile.TarInfo(name=name)
+
+        tarinfo.size = len(script)
+        tarinfo.mtime = time.time()
+        tar.addfile(tarinfo, BytesIO(script.encode("utf-8")))
+        tar.close()
+
+        return tarstream.getvalue()
+
+    def run(self, repo: str, branch: str, task: Task) -> Result:
+        self.check_docker_access()
+
+        image = self.get_or_create_environment(repo, branch)
+
+        container = self.container_running(image)
+        if container is None:
+            container = self.start_container(image)
+
+        script_name, script = self.create_script(task)
+
+        container.exec_run(["mkdir", "-p", "/srv/scripts"])
+        container.put_archive("/srv/scripts", self.tar_file(script_name, script))
+
+        try:
+            res = container.exec_run(["python", f"/srv/scripts/{script_name}"], tty=True)
+
+            return Result(json.loads(res))
+        except Exception as e:
+            logger.exception(e)
+            return Result({"success": False, "error": str(e)})
+        finally:
+            if not self.keep_container_running:
+                container.kill("9")
```

Comment:
Maybe get the number from some constant? `signal.SIGKILL` seems to be more descriptive than hardcoded 9.

Reply:
Yep, replaced